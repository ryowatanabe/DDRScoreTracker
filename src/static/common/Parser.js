import { MusicData } from './MusicData.js';
import { ScoreData } from './ScoreData.js';
import { ScoreDetail } from './ScoreDetail.js';
import { Constants } from './Constants.js';

export class Parser {
  static get STATUS() {
    return {
      SUCCESS: 0,
      UNKNOWN_ERROR: 1,
      LOGIN_REQUIRED: 2,
    };
  }

  static getResultStatus(rootElement) {
    if (rootElement.querySelector('#login.errinfo_btn') !== null) {
      return this.STATUS.LOGIN_REQUIRED;
    }
    if (rootElement.querySelector('#error') !== null) {
      return this.STATUS.UNKNOWN_ERROR;
    }
    return this.STATUS.SUCCESS;
  }

  static parseMusicList(rootElement) {
    const res = {
      hasNext: false,
      nextUrl: '',
      musics: [],
      status: this.getResultStatus(rootElement),
    };
    if (res.status != this.STATUS.SUCCESS) {
      return res;
    }
    const next = rootElement.querySelectorAll('#next.arrow');
    if (next.length > 0) {
      res.hasNext = true;
      res.nextUrl = next[0].querySelector('a').href;
    }
    const musics = rootElement.querySelectorAll('tr.data');
    musics.forEach(function (music) {
      const regexp = /^.*img=([0-9a-zA-Z]+).*$/;
      const src = music.querySelector('td img').src;
      const musicId = src.replace(regexp, '$1');
      const title = music.querySelector('.music_tit').innerHTML;
      const difficulty = Array.from(music.querySelectorAll('.difficult')).map(function (element) {
        const value = parseInt(element.innerHTML);
        return value ? value : 0;
      });
      const musicData = new MusicData(musicId, Constants.MUSIC_TYPE.NORMAL, title, difficulty);
      res.musics.push(musicData);
    });
    res.status = this.STATUS.SUCCESS;
    return res;
  }

  static parseMusicDetail(rootElement) {
    const res = {
      musics: [],
      status: this.getResultStatus(rootElement),
    };
    if (res.status != this.STATUS.SUCCESS) {
      return res;
    }
    const musicInfo = rootElement.querySelectorAll('#music_info td');
    if (musicInfo.length != 2) {
      res.status = this.STATUS.UNKNOWN_ERROR;
      return res;
    }
    const regexpForMusicId = /^.*img=([0-9a-zA-Z]+).*$/;
    const src = musicInfo[0].querySelector('img').src;
    const musicId = src.replace(regexpForMusicId, '$1');
    const title = musicInfo[1].innerHTML.split('<br>')[0];
    const regexpForDifficulties = /^.*songdetails_level_([0-9]*).png$/;
    const difficulties = Array.from(rootElement.querySelectorAll('li.step img'));
    const difficulty = difficulties.map((element) => {
      const value = parseInt(element.src.replace(regexpForDifficulties, '$1'));
      return value ? value : 0;
    });
    const musicData = new MusicData(musicId, Constants.MUSIC_TYPE.NORMAL, title, difficulty);
    res.musics.push(musicData);
    res.status = this.STATUS.SUCCESS;
    return res;
  }

  static parseScoreList(rootElement) {
    const res = {
      hasNext: false,
      nextUrl: '',
      scores: [],
      status: this.getResultStatus(rootElement),
    };
    if (res.status != this.STATUS.SUCCESS) {
      return res;
    }
    const next = rootElement.querySelectorAll('#next.arrow');
    if (next.length > 0) {
      res.hasNext = true;
      res.nextUrl = next[0].querySelector('a').href;
    }
    const isDouble = rootElement.querySelectorAll('#t_double.game_type .select').length > 0;
    const scores = Array.from(rootElement.querySelectorAll('tr.data'));
    scores.forEach(function (score) {
      const regexp = /^.*img=([0-9a-zA-Z]+).*$/;
      const src = score.querySelector('td img.jk, td img.jk2').src;
      const musicId = src.replace(regexp, '$1');
      const scoreData = new ScoreData(musicId);
      Object.keys(Constants.DIFFICULTY_NAME_MAP).forEach(function (difficultyName) {
        const difficulty = Constants.DIFFICULTY_NAME_MAP[difficultyName] + (isDouble ? Constants.DIFFICULTIES_OFFSET_FOR_DOUBLE : 0);

        const detail = score.querySelectorAll('#' + difficultyName + '.rank');
        if (detail.length == 0) {
          return;
        }

        const scoreDetail = new ScoreDetail();
        const value = parseInt(detail[0].querySelector('.data_score').innerHTML);
        scoreDetail.score = value ? value : 0;
        const regexp = /^.*\/([^\/]+)$/;
        const scoreRankFileName = detail[0].querySelectorAll('div.data_rank img')[0].src.replace(regexp, '$1');
        const clearTypeFileName = detail[0].querySelectorAll('div.data_rank img')[1].src.replace(regexp, '$1');
        scoreDetail.scoreRank = Constants.SCORE_RANK_FILE_MAP[scoreRankFileName];
        scoreDetail.clearType = Constants.CLEAR_TYPE_FILE_MAP[clearTypeFileName];

        if (scoreDetail.scoreRank == Constants.SCORE_RANK.NO_PLAY) {
          return;
        }
        scoreData.applyScoreDetail(difficulty, scoreDetail);
      });
      res.scores.push(scoreData);
    });
    res.status = this.STATUS.SUCCESS;
    return res;
  }

  static parseScoreDetail(rootElement) {
    const res = {
      scores: [],
      status: this.getResultStatus(rootElement),
    };
    if (res.status != this.STATUS.SUCCESS) {
      return res;
    }
    const detail = Array.from(rootElement.querySelectorAll('#music_detail_table td, #course_detail_table td')).map((element) => {
      return element.innerHTML;
    });
    if (detail.length > 0) {
      const musicInfo = rootElement.querySelector('#music_info td');
      const regexpForMusicId = /^.*img=([0-9a-zA-Z]+).*$/;
      const src = musicInfo.querySelector('img').src;
      const musicId = src.replace(regexpForMusicId, '$1');

      const params = new URL(document.location).searchParams;
      const difficulty = params.get('diff');

      const scoreData = new ScoreData(musicId);
      const scoreDetail = new ScoreDetail();
      scoreDetail.score = parseInt(detail[2]) ? parseInt(detail[2]) : 0;
      scoreDetail.scoreRank = Constants.SCORE_RANK_NAME_MAP[detail[1]];
      scoreDetail.clearType = Constants.CLEAR_TYPE_NAME_MAP[detail[7]];
      scoreDetail.playCount = parseInt(detail[4]) ? parseInt(detail[4]) : 0;
      scoreDetail.clearCount = parseInt(detail[8]) ? parseInt(detail[8]) : 0;
      scoreDetail.maxCombo = parseInt(detail[3]) ? parseInt(detail[3]) : 0;
      scoreData.applyScoreDetail(difficulty, scoreDetail);
      res.scores.push(scoreData);
    }
    res.status = this.STATUS.SUCCESS;
    return res;
  }
}
