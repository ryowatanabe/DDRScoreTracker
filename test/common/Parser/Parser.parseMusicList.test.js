import { Parser } from '../../../src/static/common/Parser.js';
const fs = require('fs');
const path = require('path');

test('Parser.ParseMusicList (error)', async () => {
  const html = fs.readFileSync(path.resolve(path.join(__dirname, 'fixtures/error.html')), 'utf8');
  const rootElement = document.createElement('body');
  rootElement.innerHTML = html;
  const res = Parser.parseMusicList(rootElement);
  expect(res).toMatchSnapshot();
});

test('Parser.ParseMusicList (noLogin)', async () => {
  const html = fs.readFileSync(path.resolve(path.join(__dirname, 'fixtures/no-login.html')), 'utf8');
  const rootElement = document.createElement('body');
  rootElement.innerHTML = html;
  const res = Parser.parseMusicList(rootElement);
  expect(res).toMatchSnapshot();
});

test('Parser.ParseMusicList (hasNext:true)', async () => {
  const html = fs.readFileSync(path.resolve(path.join(__dirname, 'fixtures/music-list-1.html')), 'utf8');
  const rootElement = document.createElement('body');
  rootElement.innerHTML = html;
  const res = Parser.parseMusicList(rootElement);
  expect(res).toMatchSnapshot();
});

test('Parser.ParseMusicList (hasNext:false)', async () => {
  const html = fs.readFileSync(path.resolve(path.join(__dirname, 'fixtures/music-list-17.html')), 'utf8');
  const rootElement = document.createElement('body');
  rootElement.innerHTML = html;
  const res = Parser.parseMusicList(rootElement);
  expect(res).toMatchSnapshot();
});
