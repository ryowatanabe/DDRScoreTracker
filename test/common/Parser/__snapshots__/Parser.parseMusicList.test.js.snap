// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser.ParseMusicList (error) 1`] = `
Object {
  "hasNext": false,
  "musics": Array [],
  "nextUrl": "",
  "status": 1,
}
`;

exports[`Parser.ParseMusicList (hasNext:false) 1`] = `
Object {
  "hasNext": false,
  "musics": Array [
    MusicData {
      "difficulty": Array [
        2,
        4,
        6,
        13,
        0,
        5,
        7,
        13,
        0,
      ],
      "musicId": "olloQOobddiIIPObP91l0qPl6oo8P9Dl",
      "title": "What Will Come of Me",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        6,
        9,
        11,
        0,
        6,
        7,
        11,
        0,
      ],
      "musicId": "dbq9dPPOQ60ibOol0I6b9iQol9lD669D",
      "title": "Why not",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        9,
        13,
        0,
        5,
        9,
        13,
        0,
      ],
      "musicId": "QQbiq1qQd8d1blOldi889OdQd9P1106Q",
      "title": "WH1TE RO5E",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        4,
        8,
        13,
        0,
        5,
        11,
        13,
        0,
      ],
      "musicId": "qI06bbQ1dloi0IIIO1IQ1q8bdPdl6l69",
      "title": "Wicked Plastik",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        5,
        6,
        7,
        10,
        0,
        5,
        9,
        11,
        0,
      ],
      "musicId": "01808Q1Q6lQQ1lP0qd80I0b0qqDd1OOP",
      "title": "WILD RUSH",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        11,
        13,
        6,
        9,
        11,
        13,
      ],
      "musicId": "I81iP0Io9o1l9li0Iq99P1091O8DoDOd",
      "title": "WILD SIDE",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        1,
        4,
        7,
        11,
        0,
        5,
        7,
        11,
        0,
      ],
      "musicId": "d6iI8ii9qi0Pi80dDQQlPq6qQo69860P",
      "title": "will",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        7,
        11,
        15,
        0,
        7,
        11,
        15,
        0,
      ],
      "musicId": "8bI68oIq08d99P880bDPQ9D80O099Q10",
      "title": "Windy Fairy",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        4,
        8,
        12,
        0,
        5,
        8,
        11,
        0,
      ],
      "musicId": "blI8D10Q1IDQ8Q19bi0IDlD80oD0bloi",
      "title": "Wings of an Angel (Fly With Me)",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        4,
        8,
        11,
        0,
        5,
        8,
        12,
        0,
      ],
      "musicId": "Qo1qdD0Q9d11d9Id6o09P9id8lPbOQ6o",
      "title": "Wow Wow VENUS",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        10,
        13,
        0,
        5,
        9,
        13,
        0,
      ],
      "musicId": "bDDd08iP8dlIlOo6iqd91dPiI1lQdOQq",
      "title": "xenon",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        7,
        10,
        13,
        15,
        7,
        10,
        13,
        15,
      ],
      "musicId": "PQQd6lQlDqilqliI0IobolqIDlo0Iboq",
      "title": "Xepher",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        6,
        9,
        12,
        0,
        6,
        9,
        12,
        0,
      ],
      "musicId": "8OPDQ1oQIOPdqqD18oo0lQ8OQOIbbDoP",
      "title": "Yeah! Yeah!",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        1,
        3,
        6,
        10,
        0,
        3,
        6,
        10,
        0,
      ],
      "musicId": "D0qloQo0bQliQqooQ609IP8i09old99D",
      "title": "You",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        4,
        8,
        12,
        13,
        6,
        8,
        12,
        13,
      ],
      "musicId": "0qOPDqq0qld66bll9ob9iIqiiQPiQIDQ",
      "title": "You are a Star",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        5,
        8,
        10,
        0,
        5,
        7,
        10,
        0,
      ],
      "musicId": "D8I68Ibib9ldQDIPIQdO6lqI61PQlPO9",
      "title": "You gotta move it (feat. Julie Rugaard)",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        4,
        7,
        10,
        0,
        4,
        7,
        11,
        0,
      ],
      "musicId": "oiIob81i6d6P6qbddOP6Q0d0obQ0PID9",
      "title": "Your Angel",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        11,
        15,
        0,
        7,
        11,
        15,
        0,
      ],
      "musicId": "I61ioI01IOPDPbP1OO0l98PD0iDbid0D",
      "title": "ZEPHYRANTHES",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        11,
        15,
        0,
        6,
        9,
        14,
        0,
      ],
      "musicId": "96QPqq0q6Id0Ii0QbD1bQqPbiDqb0lPo",
      "title": "ZETA～素数の世界と超越者～",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        8,
        11,
        0,
        5,
        8,
        10,
        0,
      ],
      "musicId": "IQ0dIiDql9Q68d9ddQIiQbq1OPq8Db69",
      "title": ".59",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        4,
        6,
        9,
        0,
        5,
        6,
        9,
        0,
      ],
      "musicId": "ddOOOibQ6qIi9DI1bl8b108I1Ib6QDl9",
      "title": "1998",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        4,
        7,
        8,
        0,
        3,
        6,
        8,
        0,
      ],
      "musicId": "i9Pidb18b9O8bdqQ91QQP0oPdIdIODq1",
      "title": "30 Lives (Up-Up-Down-Dance Mix)",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        5,
        9,
        11,
        0,
        5,
        8,
        11,
        0,
      ],
      "musicId": "Q96bO9D61lib19IiIi0i69P80bo6q69Q",
      "title": "321STARS",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        6,
        8,
        12,
        0,
        6,
        9,
        13,
        0,
      ],
      "musicId": "80Q1idDo6O6Db0106b0q6qbOP1P8QQb0",
      "title": "39",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        9,
        12,
        0,
        7,
        9,
        12,
        0,
      ],
      "musicId": "l1ODQ1bO9POQ6d8O0Iobbio11qb6O6bd",
      "title": "50th Memorial Songs -二人の時 ～under the cherry blossoms～-",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        5,
        10,
        14,
        0,
        5,
        10,
        14,
        0,
      ],
      "musicId": "bP10qq9dPDDqIl16lPIDDd8b00dIIP08",
      "title": "50th Memorial Songs -Beginning Story-",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        9,
        12,
        0,
        8,
        10,
        13,
        0,
      ],
      "musicId": "0IIO08bII0QQDOdobqbl8Ib8Q8O8q1li",
      "title": "50th Memorial Songs -Flagship medley-",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        5,
        9,
        13,
        0,
        5,
        9,
        13,
        0,
      ],
      "musicId": "D0iDbio81PIbI8dD6QOIP9IidDOP0PDP",
      "title": "50th Memorial Songs -The BEMANI History-",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        7,
        10,
        14,
        0,
        7,
        11,
        14,
        0,
      ],
      "musicId": "Dd98lDdo8Ob698d9qbO9iD0PqO19oQbI",
      "title": "8000000",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        13,
        14,
        18,
        7,
        12,
        15,
        18,
      ],
      "musicId": "id9oObq9P6Q6Pq6lQPqI88OP1DD8D0O1",
      "title": "888",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        5,
        11,
        14,
        17,
        18,
        10,
        14,
        17,
        18,
      ],
      "musicId": "00obPO6oPIPOoD9qb0dIl6q6D8P6o9bI",
      "title": "IX",
      "type": 0,
    },
  ],
  "nextUrl": "",
  "status": 0,
}
`;

exports[`Parser.ParseMusicList (hasNext:true) 1`] = `
Object {
  "hasNext": true,
  "musics": Array [
    MusicData {
      "difficulty": Array [
        3,
        5,
        7,
        11,
        0,
        5,
        8,
        11,
        0,
      ],
      "musicId": "QPd01OQqbOIiDoO1dbdo1IIbb60bqPdl",
      "title": "愛言葉",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        4,
        6,
        9,
        0,
        4,
        6,
        11,
        0,
      ],
      "musicId": "OboID1PloIIoOOObQdQOP110I61Ddl9I",
      "title": "蒼い衝動 ～for EXTREME～",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        5,
        9,
        13,
        0,
        5,
        10,
        13,
        0,
      ],
      "musicId": "918oPDQDoPlD1OlqlqQI08I6dD9iqoil",
      "title": "朝色の紙飛行機",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        10,
        13,
        0,
        7,
        10,
        13,
        0,
      ],
      "musicId": "16bl8D8lOddP9Obq0OIlOI6q8Pi1l9O0",
      "title": "天ノ弱",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        7,
        11,
        14,
        0,
        7,
        12,
        15,
        0,
      ],
      "musicId": "06O0ObdQobq86lPDo6P18dQ1QPdilIQO",
      "title": "妖隠し -あやかしかくし-",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        10,
        14,
        0,
        7,
        10,
        14,
        0,
      ],
      "musicId": "d0DDd1dP6llPbioqPdiQ0IDd918bDqI8",
      "title": "ありふれたせかいせいふく",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        8,
        11,
        13,
        6,
        8,
        12,
        13,
      ],
      "musicId": "8bQQ0lP96186D8Ibo8IoOd6o16qioiIo",
      "title": "アルストロメリア (walk with you remix)",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        5,
        9,
        12,
        14,
        0,
        9,
        13,
        15,
        0,
      ],
      "musicId": "1d9lDIl6iiiD0D0d8qd6dlDlq6P8o0QI",
      "title": "阿波おどり -Awaodori- やっぱり踊りはやめられない",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        1,
        4,
        8,
        12,
        0,
        5,
        9,
        12,
        0,
      ],
      "musicId": "Oq900IIl1Ob8qoIPqDP0PD6blbqIqlP1",
      "title": "いーあるふぁんくらぶ",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        8,
        12,
        0,
        4,
        8,
        11,
        0,
      ],
      "musicId": "61oIP0QIlO90d18ObDP1Dii6PoIQoOD8",
      "title": "イーディーエム・ジャンパーズ",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        5,
        8,
        11,
        0,
        3,
        7,
        11,
        0,
      ],
      "musicId": "ibI0QllidqQ089ql8dob6bIoodbbIoqq",
      "title": "怒れる大きな白い馬",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        8,
        11,
        0,
        6,
        8,
        11,
        0,
      ],
      "musicId": "16IqQ680IioldO69QD61liQod0DQ8boQ",
      "title": "色は匂へど散りぬるを",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        5,
        9,
        12,
        0,
        5,
        9,
        12,
        0,
      ],
      "musicId": "bd8oD0iqDQ8Qobd1oIqi8981QbD6I6Iq",
      "title": "エイリアンエイリアン",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        8,
        11,
        13,
        0,
        8,
        11,
        13,
        0,
      ],
      "musicId": "80PdqQ0iiOQb9i91lIliodiO9PI8O609",
      "title": "エキサイティング！！も・ちゃ・ちゃ☆",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        8,
        12,
        15,
        0,
        8,
        12,
        15,
        0,
      ],
      "musicId": "oQolO9oOO01ll0I8969i8O1IQliloDqP",
      "title": "エンドルフィン",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        8,
        12,
        16,
        0,
        8,
        12,
        16,
        0,
      ],
      "musicId": "9I00D9Id61iD6QP8i8Dd6698PoQ9bdi9",
      "title": "お米の美味しい炊き方、そしてお米を食べることによるその効果。",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        8,
        12,
        15,
        0,
        8,
        12,
        15,
        0,
      ],
      "musicId": "ilD80b8QQl9Qo0PP0lQ00b8bO9Oqoobo",
      "title": "御千手メディテーション",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        7,
        10,
        13,
        0,
        7,
        10,
        13,
        0,
      ],
      "musicId": "i91q8l0bbi8qll000ob6qDiQP891iiql",
      "title": "乙女繚乱 舞い咲き誇れ",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        1,
        3,
        8,
        11,
        0,
        3,
        8,
        11,
        0,
      ],
      "musicId": "qdQoQli1d8DP0Obb919I6P1Odilib1Dd",
      "title": "おねがいダーリン",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        8,
        11,
        13,
        6,
        10,
        12,
        14,
      ],
      "musicId": "I96dOqqqQIi9oiqbqDPbQ8I8PQbqOb1o",
      "title": "朧",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        8,
        12,
        0,
        7,
        9,
        13,
        0,
      ],
      "musicId": "oPP1Pdl6Qdlb8oq10PP1doDibdbbl9bl",
      "title": "朧 (dj TAKA Remix)",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        13,
        0,
        6,
        9,
        12,
        0,
      ],
      "musicId": "0Ib6qIb0iPOoQqqDIl90dP0PI0o9QlDd",
      "title": "女言葉の消失",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        9,
        12,
        11,
        4,
        8,
        11,
        11,
      ],
      "musicId": "l9lq19DdiD8qQPoPOlboi1qQii0IQI86",
      "title": "革命",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        10,
        13,
        0,
        5,
        9,
        13,
        0,
      ],
      "musicId": "l1lO999PqIboPQOD86PDQo1ol1iIiiOi",
      "title": "革命パッショネイト",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        4,
        7,
        9,
        0,
        4,
        6,
        9,
        0,
      ],
      "musicId": "ql1Q8P100IIlbl0Pdi08I8qD900idqQq",
      "title": "カゲロウ",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        6,
        9,
        12,
        0,
        6,
        9,
        13,
        0,
      ],
      "musicId": "id1lodDbo0Iq9088Qd19d1OlDPP01D8o",
      "title": "からふるぱすてる",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        6,
        9,
        13,
        0,
        6,
        9,
        13,
        0,
      ],
      "musicId": "i89IO1PPOl0boOol0bQI19Q9liQ0l6O1",
      "title": "キケンな果実",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        5,
        8,
        11,
        0,
        5,
        10,
        11,
        0,
      ],
      "musicId": "QIDd80o0OqobODP00ldQ1D9dl81qQi0d",
      "title": "きゅん×きゅんばっきゅん☆LOVE",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        8,
        12,
        15,
        0,
        8,
        12,
        16,
        0,
      ],
      "musicId": "QoDQ0oq08QqiI0DQ8dIPb10qIi6d6i89",
      "title": "腐れ外道とチョコレゐト",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        5,
        7,
        11,
        0,
        5,
        8,
        11,
        0,
      ],
      "musicId": "8Q8OiP8QPbl8bqID1loQq896ldPQd99o",
      "title": "倉野川音頭",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        10,
        13,
        0,
        6,
        9,
        12,
        0,
      ],
      "musicId": "q6iD8idqo69qIPl00IQidoq1o1o1d1b9",
      "title": "クリムゾンゲイト",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        10,
        14,
        0,
        6,
        10,
        14,
        0,
      ],
      "musicId": "16oDDIlP8bIODQ6Ql0881d9Qqdb19b98",
      "title": "黒髪乱れし修羅となりて～凛 edition～",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        10,
        12,
        0,
        6,
        10,
        13,
        0,
      ],
      "musicId": "IbDoODD9qb66Iqd6DbOQiOoDll1DqiO9",
      "title": "君氏危うくも近うよれ",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        13,
        0,
        5,
        10,
        12,
        0,
      ],
      "musicId": "Pd1d96Q91oiboqb86DdddD68dbq81Pdo",
      "title": "激アツ☆マジヤバ☆チアガール",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        4,
        7,
        10,
        0,
        3,
        7,
        11,
        0,
      ],
      "musicId": "qiDOD0iidOli9l0qbP6IbOD19OQ8D8Po",
      "title": "月光蝶",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        12,
        0,
        7,
        9,
        13,
        0,
      ],
      "musicId": "QlID98Ddoi6O8lP8IQd6iid610011l60",
      "title": "幻想系世界修復少女",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        9,
        12,
        0,
        7,
        9,
        13,
        0,
      ],
      "musicId": "I1qd16dPdqPi96ql0lP10OQDbo11D89P",
      "title": "恋時雨",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        7,
        12,
        14,
        17,
        8,
        13,
        15,
        17,
      ],
      "musicId": "Q9Q088PPd0bPi9QQ8iil1lQqPdiQP69D",
      "title": "恋する☆宇宙戦争っ!!",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        11,
        0,
        6,
        9,
        11,
        0,
      ],
      "musicId": "i9il1DIDl91Db0OdllOddIqO600Pl89q",
      "title": "恋のパズルマジック",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        7,
        11,
        14,
        0,
        6,
        11,
        14,
        0,
      ],
      "musicId": "qilOP8l6bDPi98lqDbO6O1obOi068QOQ",
      "title": "恋はどう？モロ◎波動OK☆方程式！！",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        12,
        15,
        6,
        9,
        12,
        15,
      ],
      "musicId": "bo9l0Dl0PbqQQQ16Qq6bOOi68o0996Pq",
      "title": "紅焔",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        11,
        0,
        6,
        9,
        11,
        0,
      ],
      "musicId": "qbbPQoqqqiPd98d61P0qDd0iDId6OqoQ",
      "title": "この青空の下で",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        7,
        10,
        14,
        0,
        7,
        10,
        14,
        0,
      ],
      "musicId": "9idOOIdDO01D8PO06bOPPbOIOoiQ8O0P",
      "title": "この子の七つのお祝いに",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        5,
        7,
        11,
        13,
        11,
        7,
        11,
        13,
        11,
      ],
      "musicId": "911lId0IooIOq1oDPoQ0iQiDiOIbd6oo",
      "title": "桜",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        4,
        9,
        12,
        0,
        5,
        9,
        12,
        0,
      ],
      "musicId": "doIiQl6oDPP6i9qobqoobblPQqPiI9bD",
      "title": "サナ・モレッテ・ネ・エンテ",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        4,
        9,
        12,
        0,
        4,
        9,
        11,
        0,
      ],
      "musicId": "1P9bqO00bIQ0q8d1Q80l0Dq8oi6boqb0",
      "title": "さよならトリップ ～夏陽 EDM edition～",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        6,
        10,
        13,
        0,
        7,
        11,
        13,
        0,
      ],
      "musicId": "D11ldIqD8dQIi9oQQIIo86QOI8D9olP8",
      "title": "幸せになれる隠しコマンドがあるらしい",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        2,
        5,
        9,
        12,
        0,
        5,
        9,
        12,
        0,
      ],
      "musicId": "oQ8oqddQOObiq0ldo68i1d6o1IidObi6",
      "title": "雫",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        4,
        8,
        11,
        13,
        0,
        7,
        11,
        13,
        0,
      ],
      "musicId": "OoQoQIP06Doq1d88dDQQdlQ8i68Do9DO",
      "title": "漆黒のスペシャルプリンセスサンデー",
      "type": 0,
    },
    MusicData {
      "difficulty": Array [
        3,
        6,
        9,
        11,
        0,
        6,
        9,
        12,
        0,
      ],
      "musicId": "0d8Q9I86lo8oDlQiPd81ll66ob0qli89",
      "title": "しゃかりきリレーション",
      "type": 0,
    },
  ],
  "nextUrl": "http://localhost/game/ddr/ddra20/p/music/index.html?offset=1&filter=0&filtertype=0&playmode=2",
  "status": 0,
}
`;

exports[`Parser.ParseMusicList (noLogin) 1`] = `
Object {
  "hasNext": false,
  "musics": Array [],
  "nextUrl": "",
  "status": 2,
}
`;
